pkg/v1/mutate/mutate.go

```diff
@@ -392,3 +386,123 @@ func inWhiteoutDir(fileMap map[string]bool, file string) bool {
 	}
 	return false
 }
+
+// Time sets all timestamps in an image to the give timestamp.
+func Time(img v1.Image, t time.Time) (v1.Image, error) {
+	newImage := empty.Image
+
+	layers, err := img.Layers()
+	if err != nil {
+		return nil, err
+	}
+
+	// Strip away all timestamps from layers
+	var newLayers []v1.Layer
+	for _, layer := range layers {
+		newLayer, err := layerTime(layer, t)
+		if err != nil {
+			return nil, err
+		}
+		newLayers = append(newLayers, newLayer)
+	}
+
+	newImage, err = AppendLayers(newImage, newLayers...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Strip away timestamps and base container info from the config file
+	cf, err := newImage.ConfigFile()
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := cf.DeepCopy()
+	cfg.Created = v1.Time{Time: t}
+
+	for _, h := range cfg.History {
+		h.Created = v1.Time{Time: t}
+	}
+
+	newImage, err = ConfigFile(newImage, cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return newImage, nil
+
+}
+
+func layerTime(layer v1.Layer, t time.Time) (v1.Layer, error) {
+	layerReader, err := layer.Uncompressed()
+	if err != nil {
+		return nil, err
+	}
+	w := new(bytes.Buffer)
+	tarWriter := tar.NewWriter(w)
+	defer tarWriter.Close()
+
+	tarReader := tar.NewReader(layerReader)
+	for {
+		header, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		header.ModTime = t
+		if err := tarWriter.WriteHeader(header); err != nil {
+			return nil, err
+		}
+
+		if header.Typeflag == tar.TypeReg {
+			_, err = io.Copy(tarWriter, tarReader)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	// gzip the contents, then create the layer
+	g, err := v1util.GzipReadCloser(ioutil.NopCloser(w))
+	if err != nil {
+		return nil, err
+	}
+
+	opener := func() (io.ReadCloser, error) {
+		return g, nil
+	}
+	layer, err = tarball.LayerFromOpener(opener)
+	if err != nil {
+		return nil, err
+	}
+
+	return layer, nil
+}
+
+//Canonical is a helper function to combine Time and ConfigFile to make a reproducible image
+func Canonical(img v1.Image) (v1.Image, error) {
+	// Set all timestamps to 0
+	created := time.Time{}
+	img, err := Time(img, created)
+	if err != nil {
+		return nil, err
+	}
+
+	cf, err := img.ConfigFile()
+	if err != nil {
+		return nil, err
+	}
+
+	// Get rid of useless random config
+	cfg := cf.DeepCopy()
+
+	cfg.Container = ""
```

Comment:
I'm not familiar enough with docker to know for sure that these are useless... in particular I'm surprised that `ContainerConfig.Hostname` isn't needed, but I'll defer to you.